ICE: `downgrade_to_delayed_bug: cannot downgrade Warning to DelayedBug: not an error`

<!--
[31mICE[0m: Rustc ./D74EF64911B83018E4513EEAFC7D7D0B9E478375DA39AE6247285289DD45575E.rs '' 'thread 'rustc' panicked at compiler/rustc_trait_selection/src/traits/specialize/mod.rs:405:17: 'downgrade_to_delayed_bug: cannot downgrade Warning to DelayedBug: not an error'', 'thread 'rustc' panicked at compiler/rustc_trait_selection/src/traits/specialize/mod.rs:405:17: 'downgrade_to_delayed_bug: cannot downgrade Warning to DelayedBug: not an error''
File: /tmp/im/D74EF64911B83018E4513EEAFC7D7D0B9E478375DA39AE6247285289DD45575E.rs
-->

auto-reduced (treereduce-rust):
````rust
trait ToUnit<'a> {
    type Unit;
}

impl<'a, T: Copy + ?Sized> ToUnit for *const T {}

trait Overlap<T> {}

type Assoc<'a, T> = <*const T as ToUnit<'a>>::Unit;

impl<T> Overlap<T> for T {}

impl<T> Overlap<for<'a> fn(&'a (), Assoc<'a, T>)> for T {}


````

original:
````rust
trait ToUnit<'a> {
    type Unit;
}

impl<'a, T: Copy + ?Sized> ToUnit<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<S<Z>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> for *const T {}

trait Overlap<T> {}

type Assoc<'a, T> = <*const T as ToUnit<'a>>::Unit;

impl<T> Overlap<T> for T {}

impl<T> Overlap<for<'a> fn(&'a (), Assoc<'a, T>)> for T {}

````

Version information
````
rustc 1.78.0-nightly (d26b41711 2024-02-13)
binary: rustc
commit-hash: d26b41711282042c4ea0c5733e7332b07cfa4933
commit-date: 2024-02-13
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 17.0.6

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0726]: implicit elided lifetime not allowed here
 --> /tmp/icemaker_global_tempdir.7EXAgSbG2HC2/rustc_testrunner_tmpdir_reporting.EW42rArCB6R1/mvce.rs:5:28
  |
5 | impl<'a, T: Copy + ?Sized> ToUnit for *const T {}
  |                            ^^^^^^ expected lifetime parameter
  |
help: indicate the anonymous lifetime
  |
5 | impl<'a, T: Copy + ?Sized> ToUnit<'_> for *const T {}
  |                                  ++++

error[E0601]: `main` function not found in crate `mvce`
  --> /tmp/icemaker_global_tempdir.7EXAgSbG2HC2/rustc_testrunner_tmpdir_reporting.EW42rArCB6R1/mvce.rs:13:59
   |
13 | impl<T> Overlap<for<'a> fn(&'a (), Assoc<'a, T>)> for T {}
   |                                                           ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.7EXAgSbG2HC2/rustc_testrunner_tmpdir_reporting.EW42rArCB6R1/mvce.rs`

thread 'rustc' panicked at compiler/rustc_trait_selection/src/traits/specialize/mod.rs:405:17:
downgrade_to_delayed_bug: cannot downgrade Warning to DelayedBug: not an error
stack backtrace:
   0:     0x7f1a29b8a476 - std::backtrace_rs::backtrace::libunwind::trace::had2a4ba5d47bdba7
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7f1a29b8a476 - std::backtrace_rs::backtrace::trace_unsynchronized::h6224624cac22e23f
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f1a29b8a476 - std::sys_common::backtrace::_print_fmt::h1cd96ea0a958eeb4
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7f1a29b8a476 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h0898038e0594ea78
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f1a29bdd030 - core::fmt::rt::Argument::fmt::h9c8c587fc2bf430b
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/core/src/fmt/rt.rs:142:9
   5:     0x7f1a29bdd030 - core::fmt::write::h680dc8fb5cd41fb3
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/core/src/fmt/mod.rs:1120:17
   6:     0x7f1a29b7dc6f - std::io::Write::write_fmt::hace8449eddf30f5f
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/io/mod.rs:1854:15
   7:     0x7f1a29b8a254 - std::sys_common::backtrace::_print::h76c75fe0c97f7198
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f1a29b8a254 - std::sys_common::backtrace::print::h5017676def85c8cc
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f1a29b8d047 - std::panicking::default_hook::{{closure}}::h9c00eba6b9c1ecc7
  10:     0x7f1a29b8cda9 - std::panicking::default_hook::hd445e5a7ac763266
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/panicking.rs:292:9
  11:     0x7f1a2c95e14c - std[d9e32580c8520915]::panicking::update_hook::<alloc[8a8bbb87e60edfc3]::boxed::Box<rustc_driver_impl[a7798918a3a3fcc7]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f1a29b8d796 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hadd1ba1309597b13
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/alloc/src/boxed.rs:2030:9
  13:     0x7f1a29b8d796 - std::panicking::rust_panic_with_hook::hb269f6fce29b14e7
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/panicking.rs:785:13
  14:     0x7f1a29b8d4e2 - std::panicking::begin_panic_handler::{{closure}}::he19e2dc3ddf4be57
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/panicking.rs:659:13
  15:     0x7f1a29b8a976 - std::sys_common::backtrace::__rust_end_short_backtrace::h78760a33e813266b
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7f1a29b8d234 - rust_begin_unwind
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/panicking.rs:647:5
  17:     0x7f1a29bd9785 - core::panicking::panic_fmt::h684cd5274df1695b
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/core/src/panicking.rs:72:14
  18:     0x7f1a2c04ad34 - <rustc_errors[cb9e4eac46489bb4]::diagnostic::Diagnostic>::downgrade_to_delayed_bug
  19:     0x7f1a2d4694ac - rustc_trait_selection[4196083c3ff476a3]::traits::specialize::report_conflicting_impls::decorate
  20:     0x7f1a2d3d62c2 - <alloc[8a8bbb87e60edfc3]::boxed::Box<dyn for<'a, 'b> core[7a10603c235cc822]::ops::function::FnOnce<(&'a mut rustc_errors[cb9e4eac46489bb4]::diagnostic_builder::DiagnosticBuilder<'b, ()>,), Output = ()>> as core[7a10603c235cc822]::ops::function::FnOnce<(&mut rustc_errors[cb9e4eac46489bb4]::diagnostic_builder::DiagnosticBuilder<()>,)>>::call_once
  21:     0x7f1a2d468d9d - rustc_trait_selection[4196083c3ff476a3]::traits::specialize::report_overlap_conflict
  22:     0x7f1a2e518ada - rustc_trait_selection[4196083c3ff476a3]::traits::specialize::specialization_graph_provider
  23:     0x7f1a2e517049 - rustc_query_impl[604b034f8c5b13b3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[604b034f8c5b13b3]::query_impl::specialization_graph_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7e92a5157dfe0e79]::query::erase::Erased<[u8; 8usize]>>
  24:     0x7f1a2dc7a321 - rustc_query_system[f55f84d2e4d6e2c7]::query::plumbing::try_execute_query::<rustc_query_impl[604b034f8c5b13b3]::DynamicConfig<rustc_query_system[f55f84d2e4d6e2c7]::query::caches::DefIdCache<rustc_middle[7e92a5157dfe0e79]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[604b034f8c5b13b3]::plumbing::QueryCtxt, false>
  25:     0x7f1a2e50bd26 - rustc_query_impl[604b034f8c5b13b3]::query_impl::specialization_graph_of::get_query_non_incr::__rust_end_short_backtrace
  26:     0x7f1a2e50c11b - rustc_hir_analysis[2e8bae333bf39439]::coherence::coherent_trait
  27:     0x7f1a2e50bec7 - rustc_query_impl[604b034f8c5b13b3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[604b034f8c5b13b3]::query_impl::coherent_trait::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7e92a5157dfe0e79]::query::erase::Erased<[u8; 1usize]>>
  28:     0x7f1a2de3189c - rustc_query_system[f55f84d2e4d6e2c7]::query::plumbing::try_execute_query::<rustc_query_impl[604b034f8c5b13b3]::DynamicConfig<rustc_query_system[f55f84d2e4d6e2c7]::query::caches::DefIdCache<rustc_middle[7e92a5157dfe0e79]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[604b034f8c5b13b3]::plumbing::QueryCtxt, false>
  29:     0x7f1a2e598855 - rustc_query_impl[604b034f8c5b13b3]::query_impl::coherent_trait::get_query_non_incr::__rust_end_short_backtrace
  30:     0x7f1a2dd55296 - rustc_hir_analysis[2e8bae333bf39439]::check_crate
  31:     0x7f1a2e5be759 - rustc_interface[26ec800ece761eb5]::passes::analysis
  32:     0x7f1a2e5be39f - rustc_query_impl[604b034f8c5b13b3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[604b034f8c5b13b3]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[7e92a5157dfe0e79]::query::erase::Erased<[u8; 1usize]>>
  33:     0x7f1a2eb2dc24 - rustc_query_system[f55f84d2e4d6e2c7]::query::plumbing::try_execute_query::<rustc_query_impl[604b034f8c5b13b3]::DynamicConfig<rustc_query_system[f55f84d2e4d6e2c7]::query::caches::SingleCache<rustc_middle[7e92a5157dfe0e79]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[604b034f8c5b13b3]::plumbing::QueryCtxt, false>
  34:     0x7f1a2eb2d987 - rustc_query_impl[604b034f8c5b13b3]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  35:     0x7f1a2e94be3b - rustc_interface[26ec800ece761eb5]::interface::run_compiler::<core[7a10603c235cc822]::result::Result<(), rustc_span[937fc3ba6a1f0b5c]::ErrorGuaranteed>, rustc_driver_impl[a7798918a3a3fcc7]::run_compiler::{closure#0}>::{closure#0}
  36:     0x7f1a2eb84512 - std[d9e32580c8520915]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[26ec800ece761eb5]::util::run_in_thread_with_globals<rustc_interface[26ec800ece761eb5]::util::run_in_thread_pool_with_globals<rustc_interface[26ec800ece761eb5]::interface::run_compiler<core[7a10603c235cc822]::result::Result<(), rustc_span[937fc3ba6a1f0b5c]::ErrorGuaranteed>, rustc_driver_impl[a7798918a3a3fcc7]::run_compiler::{closure#0}>::{closure#0}, core[7a10603c235cc822]::result::Result<(), rustc_span[937fc3ba6a1f0b5c]::ErrorGuaranteed>>::{closure#0}, core[7a10603c235cc822]::result::Result<(), rustc_span[937fc3ba6a1f0b5c]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[7a10603c235cc822]::result::Result<(), rustc_span[937fc3ba6a1f0b5c]::ErrorGuaranteed>>
  37:     0x7f1a2eb8433e - <<std[d9e32580c8520915]::thread::Builder>::spawn_unchecked_<rustc_interface[26ec800ece761eb5]::util::run_in_thread_with_globals<rustc_interface[26ec800ece761eb5]::util::run_in_thread_pool_with_globals<rustc_interface[26ec800ece761eb5]::interface::run_compiler<core[7a10603c235cc822]::result::Result<(), rustc_span[937fc3ba6a1f0b5c]::ErrorGuaranteed>, rustc_driver_impl[a7798918a3a3fcc7]::run_compiler::{closure#0}>::{closure#0}, core[7a10603c235cc822]::result::Result<(), rustc_span[937fc3ba6a1f0b5c]::ErrorGuaranteed>>::{closure#0}, core[7a10603c235cc822]::result::Result<(), rustc_span[937fc3ba6a1f0b5c]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[7a10603c235cc822]::result::Result<(), rustc_span[937fc3ba6a1f0b5c]::ErrorGuaranteed>>::{closure#1} as core[7a10603c235cc822]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x7f1a29b96985 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h75639a97b430af13
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/alloc/src/boxed.rs:2016:9
  39:     0x7f1a29b96985 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h13d2154aa610cd72
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/alloc/src/boxed.rs:2016:9
  40:     0x7f1a29b96985 - std::sys::pal::unix::thread::Thread::new::thread_start::h77fba1514fb25e78
                               at /rustc/d26b41711282042c4ea0c5733e7332b07cfa4933/library/std/src/sys/pal/unix/thread.rs:108:17
  41:     0x7f1a299809eb - <unknown>
  42:     0x7f1a29a047cc - <unknown>
  43:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.78.0-nightly (d26b41711 2024-02-13) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [specialization_graph_of] building specialization graph of trait `Overlap`
#1 [coherent_trait] coherence checking all impls of trait `Overlap`
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0601, E0726.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
#0 [specialization_graph_of] building specialization graph of trait `Overlap`
#1 [coherent_trait] coherence checking all impls of trait `Overlap`
#2 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_trait_selection/src/traits/specialize/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/specialize/mod.rs
@@ -402,10 +402,6 @@ fn report_conflicting_impls<'tcx>(
         impl_span: Span,
         err: &mut Diag<'_, G>,
     ) {
-        if (overlap.trait_ref, overlap.self_ty).references_error() {
-            err.downgrade_to_delayed_bug();
-        }
-
         match tcx.span_of_impl(overlap.with_impl) {
             Ok(span) => {
                 err.span_label(span, "first implementation here");
@@ -457,6 +453,11 @@ fn report_conflicting_impls<'tcx>(
             }
         )
     });
+
+    // Don't report overlap errors if the header references error
+    if let Err(err) = (overlap.trait_ref, overlap.self_ty).error_reported() {
+        return Err(err);
+    }
 
     match used_to_be_allowed {
         None => {

--- a/tests/ui/coherence/skip-reporting-if-references-err.current.stderr
+++ b/tests/ui/coherence/skip-reporting-if-references-err.current.stderr
@@ -1,0 +1,27 @@
+error[E0726]: implicit elided lifetime not allowed here
+  --> $DIR/skip-reporting-if-references-err.rs:10:9
+   |
+LL | impl<T> ToUnit for T {}
+   |         ^^^^^^ expected lifetime parameter
+   |
+help: indicate the anonymous lifetime
+   |
+LL | impl<T> ToUnit<'_> for T {}
+   |               ++++
+
+error[E0277]: the trait bound `for<'a> (): ToUnit<'a>` is not satisfied
+  --> $DIR/skip-reporting-if-references-err.rs:15:29
+   |
+LL | impl Overlap for for<'a> fn(<() as ToUnit<'a>>::Unit) {}
+   |                             ^^^^^^^^^^^^^^^^^^^^^^^^ the trait `for<'a> ToUnit<'a>` is not implemented for `()`
+
+error[E0277]: the trait bound `for<'a> (): ToUnit<'a>` is not satisfied
+  --> $DIR/skip-reporting-if-references-err.rs:15:18
+   |
+LL | impl Overlap for for<'a> fn(<() as ToUnit<'a>>::Unit) {}
+   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `for<'a> ToUnit<'a>` is not implemented for `()`
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0277, E0726.
+For more information about an error, try `rustc --explain E0277`.

--- a/tests/ui/coherence/skip-reporting-if-references-err.next.stderr
+++ b/tests/ui/coherence/skip-reporting-if-references-err.next.stderr
@@ -1,0 +1,14 @@
+error[E0726]: implicit elided lifetime not allowed here
+  --> $DIR/skip-reporting-if-references-err.rs:10:9
+   |
+LL | impl<T> ToUnit for T {}
+   |         ^^^^^^ expected lifetime parameter
+   |
+help: indicate the anonymous lifetime
+   |
+LL | impl<T> ToUnit<'_> for T {}
+   |               ++++
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0726`.

--- a/tests/ui/coherence/skip-reporting-if-references-err.rs
+++ b/tests/ui/coherence/skip-reporting-if-references-err.rs
@@ -1,0 +1,19 @@
+// Regression test for #121006.
+//@ revisions: current next
+//@ ignore-compare-mode-next-solver (explicit revisions)
+//@[next] compile-flags: -Znext-solver
+
+trait ToUnit<'a> {
+    type Unit;
+}
+
+impl<T> ToUnit for T {}
+//~^ ERROR implicit elided lifetime not allowed here
+
+trait Overlap {}
+impl<U> Overlap for fn(U) {}
+impl Overlap for for<'a> fn(<() as ToUnit<'a>>::Unit) {}
+//[current]~^ ERROR the trait bound `for<'a> (): ToUnit<'a>` is not satisfied
+//[current]~| ERROR the trait bound `for<'a> (): ToUnit<'a>` is not satisfied
+
+fn main() {}

